lib/matplotlib/colorbar.py

```diff
@@ -294,6 +228,68 @@ def draw(self, renderer):
         return ret
 
 
+class _ColorbarAxesLocator:
+    """
+    Wrap any locator on the axes, or its position, to shrink the
+    inner axes if there are extend triangles at either min or max.
+    """
+    def __init__(self, cbar):
+        """
+        *bounds* (a ``[l, b, w, h]`` rectangle) and *transform* together
+        specify the position of the inset axes.
+        """
+        self._cbar = cbar
+        self._orig_locator = cbar.ax._axes_locator
+
+    def __call__(self, ax, renderer):
+
+        # make sure that lims and scales are the same
```

Comment:
I'm not sure why this is necessary; I understand from some comments that this is due to aspect ratio handling?

Reply:
Right, the aspect of the axes is presently set via the usual `ax.set_aspect`, but that requires the limits and scales of the short and long axes to be the same.  

*However* I'm starting to think that is a mistake, and we should perhaps just use something like `set_box_aspect` and get around all this awkwardness. Or just do it with the new colorbar axes locator.

Reply:
See comments, this has all be re-engineered.